src/clr-angular/data/datagrid/providers/columns.service.ts

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import { Injectable, OnDestroy } from '@angular/core';
+import { BehaviorSubject, Subscription } from 'rxjs';
+import { DatagridColumnState } from '../interfaces/column-state.interface';
+import { DatagridRenderStep } from '../enums/render-step.enum';
+import { DatagridRenderOrganizer } from '../render/render-organizer';
+
+@Injectable()
+export class ColumnsService implements OnDestroy {
```

Comment:
@gnomeontherun IMO, we are making it too complex. I was imagining a central state mangager class to be much simpler as holding properties for a very few things such as an array of column state objects and observable that any column, cell or whatever that need to subscribe to in order to get state changes. 

Currently, this class holds `BehaviorSubject` for each columns state, and also looks like it's trying to manage the subscriptions in itself too. Also, any column or cell that tries to subsribe to the state changes needs to instantiate another `BehaviorSubject` as it looks in `header-renderer.ts` as well as `cell-renderer.ts`. 

My questions is, why did we need to take this route instead of the way I mentioned in the beginning?

Reply:
The idea is that a column is a thing that has state but that state matters to all of the cells in . that column, and only cells in that column need to know the current state of that column. There are only as many `BehaviorSubject`s as there are columns, and when the list of columns changes the corresponding column observable is passed to the cells (and subscribed to at the cell level). If there are 101 cells (header and 100 rows) in a column, they all subscribe to the same subject and it becomes simple to alert to a single column about a state change. For width changes from resizing, we only need to change the affected column instead of recompute the whole datagrid.

Having just a central service with an array of widths, an array of orderings, and so forth does not keep the column state as a single unit (which is basically what the current render organizer was doing for widths). 

Maybe I'm not clear on your question, if so we can talk about it separately.

Reply:
This approach makes sense to me. It collects the column states into one place. It also allows the main renderer to create/init the states when it it knows everything is ready for it. 

It seems like a good approach to have cells/headers subscribe to the changes I wasn't sure about exposing the BehaviorSubjects directly as commented.